app/services/themes/theme_site_setting_manager.rb

```diff
@@ -0,0 +1,140 @@
+  include Service::Base
+
+  params do
+    attribute :theme_id, :integer
+    attribute :name
+    attribute :value, :string
+
+    validates :theme_id, presence: true
+    validates :name, presence: true
+
+    after_validation do
+      self.name = self.name.to_sym if self.name.present?
+      self.value = self.value || nil
+    end
+  end
+
+  policy :current_user_is_admin
+  policy :ensure_setting_is_themeable
+  model :theme
+  model :existing_theme_site_setting, optional: true
+
+  transaction do
+    step :convert_new_value_to_site_setting_values
+    step :save_update_or_destroy
+    step :log_change
+  end
+
+  step :update_site_setting_cache
+
+  private
+
+  def current_user_is_admin(guardian:)
+    guardian.is_admin?
+  end
+
+  def ensure_setting_is_themeable(params:)
+    SiteSetting.themeable[params.name]
+  end
+
+  def fetch_theme(params:)
+    Theme.find_by(id: params.theme_id)
+  end
+
+  def fetch_existing_theme_site_setting(params:, theme:)
+    theme.theme_site_settings.find_by(name: params.name)
+  end
+
+  def convert_new_value_to_site_setting_values(params:)
+    if params.value.nil?
+      context[:setting_db_value] = nil
+      context[:setting_data_type] = nil
+      context[:setting_ruby_value] = nil
+      return
+    end
+
+    # This must be done because we want the schema and data of ThemeSiteSetting to reflect
+    # that of SiteSetting, since they are the same data types and values.
+    setting_db_value, setting_data_type =
+      SiteSetting.type_supervisor.to_db_value(params.name, params.value)
+    setting_ruby_value =
+      SiteSetting.type_supervisor.to_rb_value(params.name, params.value, setting_data_type)
+
+    context[:setting_db_value] = setting_db_value
+    context[:setting_data_type] = setting_data_type
+    context[:setting_ruby_value] = setting_ruby_value
+  end
+
+  def save_update_or_destroy(
+    params:,
+    existing_theme_site_setting:,
+    theme:,
+    setting_db_value:,
+    setting_data_type:,
+    setting_ruby_value:
+  )
+    setting_record = nil
+    context[:previous_value] = nil
+    context[:new_value] = nil
+
+    if existing_theme_site_setting
+      context[:previous_value] = existing_theme_site_setting.setting_rb_value
+
+      # Since the site setting itself doesn't matter, if we are
+      # setting this back to the same value as the default setting
+      # value then it makes sense to get rid of the theme site setting
+      # override.
+      if params.value.nil? || setting_ruby_value == SiteSetting.defaults[params.name]
+        context[:new_value] = SiteSetting.defaults[params.name]
+        existing_theme_site_setting.destroy!
+      else
+        context[:new_value] = setting_ruby_value
+        existing_theme_site_setting.update!(value: setting_db_value)
+        setting_record = existing_theme_site_setting
+      end
+    else
+      # Don't need to create a record if the value is nil or matches the
+      # site setting default
```

Comment:
What if:

1. Core default for `enable_welcome_banner` is `false`
2. Theme specifies `enable_welcome_banner: false` in `about.json`, and is installed on a site. This matches the core default, so no ThemeSiteSetting record is created

3. Core changes default for `enable_welcome_banner` to `false`

4. `enable_welcome_banner` will be resolved to `false`, even though the theme specifies `true` in about.json

Perhaps we should create the record anyway, so that all the information is in the database, and we can resolve things when serializing to the client?

Reply:
Okay yes I  think this makes sense (though there are some true/false typos here). I'm not sure why I didn't create records if we match default originally -- perhaps just trying to insert unnecessary rows. But you're right that this doesn't take into account the "shifting  sands"   of changing site setting defaults.

Will update this logic for both the UI and the remote theme importer.